Jetpack Pro Dashboard: implement monitor notification email resend code

yashwin opened the pull request:
Related to 1204408201748644-as-1204666762460286

## Proposed Changes

This PR implements resend code for the monitor notification email.

#### Testing Instructions

**Prerequisites**

Since these changes are made specifically for agencies, you must set yourself(partner) as an agency - 2c49b-pb. Make sure to switch it back to the previous type.

**Instructions**

1. Apply this D111846-code and sandbox the public API.
2. Run `git checkout add/implement-monitor-resend-code` and `yarn start-jetpack-cloud`
3. Open http://jetpack.cloud.localhost:3000/, and you'll be redirected to the /dashboard.
4. Enable monitor if not enabled already.
5. Click on the clock icon next to the monitor toggle.
6. Click the `+ Add email address` button -> Enter the name and email `testemail@email.com` -> Click `Verify`.
7.  Enter a random(wrong) verification code and verify the below error is shown.

<img width="472" alt="Screenshot 2023-06-02 at 3 21 25 PM" src="https://github.com/Automattic/wp-calypso/assets/10586875/af0f6a83-91df-4e7e-9c2a-49430222935f">

8. Click on `resend` and verify the button is disabled when the request is being made.

<img width="473" alt="Screenshot 2023-06-02 at 3 22 27 PM" src="https://github.com/Automattic/wp-calypso/assets/10586875/9a35a9dc-6fe5-4b18-8909-df1934e1df51">

<img width="474" alt="Screenshot 2023-06-06 at 5 06 00 PM" src="https://github.com/Automattic/wp-calypso/assets/10586875/8006bc44-2428-4baf-81e3-abc40f9bca2f">


9. Verify the help text looks like the below once the resend request is successful. 

<img width="470" alt="Screenshot 2023-06-02 at 2 53 44 PM" src="https://github.com/Automattic/wp-calypso/assets/10586875/502971e2-c062-4bb6-b224-9311639684f4">

10. Follow the instruction - D111846-code on how to get the verification code, enter the verification code and click `Verify` and make sure the email address is verified. 

## Pre-merge Checklist

<!--
Complete applicable items on this checklist **before** merging into trunk. Inapplicable items can be left unchecked.

Both the PR author and reviewer are responsible for ensuring the checklist is completed.
-->

- [x] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [ ] [Have you written new tests](https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md) for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [x] Have you checked for TypeScript, React or other console errors?
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
- [ ] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-ajp-p2)?

yashwin committed:
Implement resend code for contact verification

Changes

Fetch on window focus & remove stale time

jkguidaven commented on client/jetpack-cloud/sections/agency-dashboard/sites-overview/style.scss:
```diff
@@ -455,3 +455,8 @@
+
+.is-loading {
+	opacity: 0.5;
+	pointer-events: none;
```
Is the `disabled` prop in the Button component already preventing click events?

yashwin replied:
Nice catch. I have removed it -https://github.com/Automattic/wp-calypso/pull/77740/commits/ebf4651a772616860fe33bf5014fb83e38087561

jkguidaven commented on client/jetpack-cloud/sections/agency-dashboard/downtime-monitoring/configure-email-notification/email-address-editor.tsx:
```diff
@@ -61,13 +68,43 @@ export default function EmailAddressEditor( {
+	}, [ emailItem.email ] );
+
+	// Function to handle resend code button click
+	const handleResendCodeClick = useCallback( () => {
```
Maybe we can merge `handleResendCode` and `handleResendCodeClick` as one function?

yashwin replied:
Since we are using resend-verification logic when the user clicks on `Verify` or `Pending` after they have decided to verify "Later", we don't want to show the message in the UI when the resend button is clicked. Plus, I would like to record the event only when the `resend` button is clicked. This is why I separated the logic here.

jkguidaven replied:
Thanks for clarifying this. make sense to me. 👍

jkguidaven commented on client/data/agency-dashboard/use-fetch-monitor-verified-contacts.ts:
```diff
@@ -24,8 +24,6 @@ const useFetchMonitorVerfiedContacts = ( isPartnerOAuthTokenLoaded: boolean ) =>
 				};
 			},
 			enabled: isPartnerOAuthTokenLoaded && isMultipleEmailEnabled,
-			refetchOnWindowFocus: false,
```
I'm curious why we change this.

yashwin replied:
This was not required. I would always like to keep the contacts updated.

jkguidaven replied:
Thanks for explaining this.  I kind of wonder why we set this `false` before and assumed it was intended. But yes, I agree we would want to have updated contacts. 👍

jkguidaven commented on client/jetpack-cloud/sections/agency-dashboard/downtime-monitoring/notification-settings/index.tsx:
```diff
@@ -81,15 +84,17 @@ export default function NotificationSettings( {
-				JSON.stringify(
-					initialSettings?.emailContacts?.map( ( { name, email } ) => ( { name, email } ) )
-				) );
+			mapAndStringifyEmails( allEmailItems ) !==
```
Maybe we can utilize Lodash's `isEqual` here. If not you can ignore my comment.

yashwin replied:
I think there were discussions on getting rid of `Lodash`. So, I would avoid using it.

jkguidaven replied:
I didn't know that. I'll keep this in mind. Thanks for sharing this. 👍

jkguidaven commented on client/jetpack-cloud/sections/agency-dashboard/downtime-monitoring/notification-settings/index.tsx:
```diff
@@ -137,10 +142,11 @@ export default function NotificationSettings( {
 			const extraEmails = allEmailItems.filter( ( item ) => ! item.isDefault );
 			params.contacts = {
 				emails: extraEmails.map( ( item ) => {
+					const isVerified = verifiedContacts.emails.includes( item.email ) || item.verified;
```
```
const isVerified = item.verified || verifiedContacts.emails.includes( item.email );
```
This way, we do not iterate if the item is verified to avoid more expensive operation.

yashwin replied:
Good suggestion! I have made the changes here - https://github.com/Automattic/wp-calypso/pull/77740/commits/ebf4651a772616860fe33bf5014fb83e38087561

jkguidaven commented on client/jetpack-cloud/sections/agency-dashboard/downtime-monitoring/configure-email-notification/email-address-editor.tsx:
```diff
@@ -61,13 +68,43 @@ export default function EmailAddressEditor( {
+	// Function to handle resend code button click
+	const handleResendCodeClick = useCallback( () => {
+		setHelpText( undefined );
+		setResendCodeClicked( true );
```
Can we also clear the verification code input field when we press resend code?

yashwin replied:
Totally! I have made the changes - https://github.com/Automattic/wp-calypso/pull/77740/commits/ebf4651a772616860fe33bf5014fb83e38087561

jkguidaven reviewed (commented):
Working as expected as per the instructions. nice work @yashwin 

I left some few comments. 👍

jkguidaven reviewed (commented):
Working as expected as per the instructions. nice work @yashwin 

I left some few comments. 👍

jkguidaven commented:
I also would like to suggest showing some text indicating we are resending the code when it is submitting. There is a brief **Please wait for a minute. If you didn’t receive it, we can resend it.** text that displays when it is loading which is not really appropriate. 

https://github.com/Automattic/wp-calypso/assets/56598660/92318f85-c41c-49de-9c3e-3a672e1339d7

yashwin committed:
Address PR comments

yashwin commented:
@jkguidaven Thanks for your review and some good suggestions. I have addressed all your comments.

> I also would like to suggest showing some text indicating we are resending the code when it is submitting. There is a brief Please wait for a minute. If you didn’t receive it, we can resend it. text that displays when it is loading which is not really appropriate.

Agree. I have made this change.

<img width="474" alt="Screenshot 2023-06-06 at 5 06 00 PM" src="https://github.com/Automattic/wp-calypso/assets/10586875/8006bc44-2428-4baf-81e3-abc40f9bca2f">

jkguidaven reviewed (approved):
Thanks for addressing my comments @yashwin. Retested the changes and it is working as expected.

👍 good job

a8ci18n commented:
This Pull Request is now available for translation here: https://translate.wordpress.com/deliverables/8027704

Thank you @yashwin for including a screenshot in the description! This is really helpful for our translators.

a8ci18n commented:
Translation for this Pull Request [has now been finished](https://translate.wordpress.com/deliverables/overview/8027704).
